Plug-Ins/Menu Editor/MenuEditorWindowController.swift

```diff
@@ -0,0 +1,395 @@
+            if markCharacter != 0 {
+                newItem.markCharacter = String(data: Data([markCharacter]), encoding: .macOSRoman) ?? ""
+            }
+
+            itemEnableBitIndex += 1
+            
+            switch commandsSize {
+            case .int16:
+                if (reader.bytesRead % 2) != 0 {
+                    try reader.advance(1)
+                }
+                let shortCommand: UInt16 = try reader.read()
+                newItem.menuCommand = UInt32(shortCommand)
+            case .int32:
+                if (reader.bytesRead % 2) != 0 {
+                    try reader.advance(1)
+                }
+                let longCommand: UInt32 = try reader.read()
+                newItem.menuCommand = longCommand
+            case .none:
+                break // only breaks out of switch
+            }
+            
+            newMenu.items.append(newItem)
+        }
+        try reader.advance(1)
+        
+        return newMenu
+    }
+    
+    /// Parse the resource into our ``items`` list.
+    private func loadItems() {
+        if resource.data.isEmpty {
+            createEmptyResource()
+        }
+        do {
+            menuInfo = try itemsFromData(resource.data)
+        } catch {
+            menuInfo = Menu()
+            self.window?.presentError(error)
+        }
+    }
+    
+    /// Create a valid but empty Menu resource. Used when we are opened for an empty resource.
+    private func createEmptyResource() {
+        let writer = BinaryDataWriter()
+        writer.write(Int16(resource.id))
+        writer.write(Int16(0)) // width
+        writer.write(Int16(0)) // height
+        writer.write(Int16(0)) // mdef ID
+        writer.write(Int16(0)) // filler
+        writer.write(UInt32.max) // enableFlags
+        let newName = (resource.name.isEmpty ? NSLocalizedString("New Menu", comment: "") : resource.name)
+        try! writer.writePString(newName) // menu title
+        writer.write(UInt8(0)) // zero terminator
+        resource.data = writer.data
+        
+        self.setDocumentEdited(true)
+    }
+    
+    private func currentResourceStateAsData() throws -> Data {
+        let writer = BinaryDataWriter()
+        
+        let commandsSize: MenuItem.CommandsSize
+        if resource.typeCode == "cmnu" {
+            commandsSize = .int16
+        } else if resource.typeCode == "CMNU" {
+            commandsSize = .int32
+        } else {
+            commandsSize = .none
+        }
+        
+        writer.write(menuInfo.menuID)
+        writer.write(Int16(0)) // width
+        writer.write(Int16(0)) // height
+        writer.write(menuInfo.mdefID) // mdef ID
+        writer.write(Int16(0)) // filler
+        writer.write(menuInfo.enableFlags) // enableFlags
+        try writer.writePString(menuInfo.name)
+        for item in menuInfo.items {
+            try writer.writePString(item.name)
+            writer.write((item.iconID == 0) ? Int8(0) : Int8(item.iconID - 256))
```

Comment:
This is an easy one to crash on if you enter a simple number like 1. Perhaps you can do a NumberFormatter with range 128-511 and allow blank for the zero case?

Reply:
257…511 ? Sure.

Reply:
Er, yeah that sounds better. But you'll need to use a UInt8 - the Int8 is still crashing.

Reply:
Derp.